pyro/infer/rws.py

```diff
@@ -0,0 +1,242 @@
+            log_weights = log_joints - log_qs.detach()
+
+            # compute wake theta loss
+            log_sum_weight = torch.logsumexp(log_weights, dim=0)
+            wake_theta_loss = -(log_sum_weight - math.log(self.num_particles)).sum()
+            warn_if_nan(wake_theta_loss, "wake theta loss")
+
+        if self.insomnia > 0:
+            # compute wake phi loss
+            normalised_weights = (log_weights - log_sum_weight).exp().detach()
+            wake_phi_loss = -(normalised_weights * log_qs).sum()
+            warn_if_nan(wake_phi_loss, "wake phi loss")
+
+        if self.insomnia < 1:
+            # compute sleep phi loss
+            _model = pyro.poutine.uncondition(model)
+            _guide = guide
+            _log_q = 0.
+
+            if self.vectorize_particles:
+                if self.max_plate_nesting == float('inf'):
+                    self._guess_max_plate_nesting(_model, _guide, *args, **kwargs)
+                _model = self._vectorized_num_sleep_particles(_model)
+                _guide = self._vectorized_num_sleep_particles(guide)
+
+            for _ in range(1 if self.vectorize_particles else self.num_sleep_particles):
+                _model_trace = poutine.trace(_model).get_trace(*args, **kwargs).detach()
+                _guide_trace = self._get_matched_trace(_model_trace, _guide, *args, **kwargs)
+                _log_q += _guide_trace.log_prob_sum()
+
+            sleep_phi_loss = -_log_q / self.num_sleep_particles
+            warn_if_nan(sleep_phi_loss, "sleep phi loss")
+
+        # compute phi loss
+        phi_loss = sleep_phi_loss if self.insomnia == 0 \
+            else wake_phi_loss if self.insomnia == 1 \
+            else self.insomnia * wake_phi_loss + (1. - self.insomnia) * sleep_phi_loss
+
+        return wake_theta_loss, phi_loss
+
+    def loss(self, model, guide, *args, **kwargs):
+        """
+        :returns: returns model loss and guide loss
+        :rtype: float, float
+
+        Computes the re-weighted wake-sleep estimators for the model (wake-theta) and the
+          guide (insomnia * wake-phi + (1 - insomnia) * sleep-phi).
+        """
+        with torch.no_grad():
+            wake_theta_loss, phi_loss = self._loss(model, guide, *args, **kwargs)
+
+        return wake_theta_loss, phi_loss
+
+    def loss_and_grads(self, model, guide, *args, **kwargs):
+        """
+        :returns: returns model loss and guide loss
+        :rtype: float
+
+        Computes the RWS estimators for the model (wake-theta) and the guide (wake-phi).
+        Performs backward as appropriate on both, using num_particle many samples/particles.
+        """
+        wake_theta_loss, phi_loss = self._loss(model, guide, *args, **kwargs)
+        try:
+            wake_theta_loss.backward(retain_graph=True)
+        except RuntimeError:
+            pass
+        phi_loss.backward()
+
+        return wake_theta_loss.detach(), phi_loss.detach()
```

Comment:
nit: I believe all losses currently return floats. Could you 
```py
return ....detach().item(), ....detach().item()
```
here?

Reply:
Done in 7d7b5f16